### PR TITLE
ENH: Boundary conditions for CubicSpline

### PR DESCRIPTION
**Pull request by nmayorov**
I decided to finish #5734 in order to progress towards 0.18.0.

The main improvement relative to the old PR is a test coverage. I believe it's very good know, all possible combinations of boundary conditions are tested as well as corner cases (n=2 and n=3).

Minor changes in logic are introduced. None boundary condition is eliminated.

@ev-br @matmel, help with a review.

### CODE REVIEW

**Comment by ev-br** [comment]
My main complaint here is still that `bc_type="periodic"` does not produce something which is actually periodic when evaluated:

```
In [18]: theta = 2 * np.pi * np.linspace(0, 1, 5)

In [19]: y = np.c_[np.cos(theta), np.sin(theta)]

In [20]: cs = CubicSpline(theta, y, bc_type='periodic')

In [21]: cs(2)
Out[21]: array([-0.39965931,  0.89821024])

In [22]: cs(2.*np.pi + 2)
Out[22]: array([-2.46375751,  0.87781022])
```

File: scipy/interpolate/_cubic.py [hunk 1]
```diff
@@ -491,34 +570,177 @@ def __init__(self, x, y, axis=0, extrapolate=True):
...
+                b2 = np.zeros_like(b1)
+                b2[0] = -a_0_m1
+                b2[-1] = -a_m2_m1
+
+                # s1 and s2 are the solutions of (n-2, n-2) system
+                s1 = solve_banded((1, 1), Ac, b1, overwrite_ab=False,
+                                  overwrite_b=False, check_finite=False)
+
+                s2 = solve_banded((1, 1), Ac, b2, overwrite_ab=False,
+                                  overwrite_b=False, check_finite=False)
```
Comment (ev-br) [inline scipy/interpolate/_cubic.py]: (Non-binding comment) these can be packed into a single call to `solve_banded` with r.h.s. with shape (N, 2).

**Comment by nmayorov** [comment]
I believe this part should be implemented as `PPoly` enhancement, say with `extrapolate='periodic'`. Does it make sense to consider it in scope of another pull request?

**Comment by josef-pkt** [comment]
@ev-br 

Isn't it periodic modulo domain?
The main application in stats for periodic (or whatever is the standard) name is to get circular connectedness in value and derivative, but the underlying data is modulo season, e.g. hour of a day or month of the year for daily or annual seasonality.

File: scipy/interpolate/_cubic.py (hunk 1 above)
Comment (nmayorov) [inline scipy/interpolate/_cubic.py]: `b1` and `b2` may have arbitrary number of dimensions, it will get tricky to handle this. I think it's better to leave @matmel's code intact.

**Comment by ev-br** [comment]
@nmayorov Sure, this could/should be taken care of in a separate PR. However the current behavior is I think surprising in a bad way. IMO we should at least document it and give a recipe.

@josef-pkt Maybe I'm dense, I don't follow. Could you elaborate?

**Comment by ev-br** [comment]
@josef-pkt Continuing from the 2nd example in the docstring:

```
In [56]: xs = np.linspace(0, 2.*np.pi, 101)

In [57]: xs1 = np.linspace(0, 2.6*np.pi, 101)

In [58]: plt.plot(cs(xs)[:, 0], cs(xs)[:, 1], '-', lw=7, alpha=0.7)
Out[58]: [<matplotlib.lines.Line2D at 0x7f93e5ddce80>]

In [59]: plt.plot(cs(xs1)[:, 0], cs(xs1)[:, 1], 'g-', lw=2)
Out[59]: [<matplotlib.lines.Line2D at 0x7f93e5ddceb8>]
```

generates

figure_1-2

**Comment by nmayorov** [comment]
I added a note about periodic evaluation. Now `extrapolate=None` by default and set to False for periodic conditions.

This simple recipe from Notes should be transferred to `PPoly` and `BPoly` eval methods.

**Comment by josef-pkt** [comment]
@ev-br  I didn't realize that it extrapolates by default. Also I'm thinking more of smoothing splines than interpolation.
However, in your example, I would never evaluate the function outside of the initial support [0, 2*pi].
I would always map the coordinates back to the initial domain [0, 2*pi] or whatever.

example: daylight brightness or whatever with a strong daily cycle assuming no seasonal effect
If we use datetime, then time is linear and we have periodic observations with many cycles. However datetime doesn't include the hour information specifically. 
If we define our explanatory time variable as hour of the day, then it is restricted to the domain [0, 24). The important part of periodic splines is that we get continuity and differentiability between before and after midnight

It's a bit similar to the discussion on circular distribution. The question is who does the `modulo` operation to map into the relevant (linear) domain, user or library function.

**Comment by nmayorov** [comment]
As you agree about implementing `extrapolate='periodic'` I can send a PR. To my taste it is a very logical and convenient option to have.

**Comment by josef-pkt** [comment]
A question on the length of a cycle/season, given that I haven't looked at the code:

What happens if there is a gap in the observations at the boundary?
In the 24 hour example, suppose the first observation is at a minute after midnight and the last information is a minute before midnight. So the observed cycle length is 24 hours minus two minutes. Do we get periodicity at 24 hours? Or, is the day a bit short and will shift in long run extrapolation?

**Comment by nmayorov** [comment]
@josef-pkt, it is implemented as follows (actually done by @matmel): 

Values `y` at `x[0]` and `x[-1]` must be `allclose` with `rtol=atol=1e-15`, otherwise an error is raised. The period is naturally assumed to be `x[-1] - x[0]`. 

If you want to see other logic, it's possible to consider it.

**Comment by nmayorov** [comment]
But I want to point out, that this is a classical formulation for periodic splines and it's better to keep it.

**Comment by josef-pkt** [comment]
> Values y at x[0] and x[-1] must be allclose with rtol=atol=1e-15, otherwise an error is raised. The period is naturally assumed to be x[-1] - x[0].

The first part sounds ok because we can duplicate the boundary point to both sides. AFAICS, in that case we also get the right period length, although maybe slightly offset from midnight like  `[0.01, 24.01]`.

(For smoothing spline estimation it's a bit different because we can choose knots for the basis functions that don't correspond to observation points.)

**Comment by nmayorov** [comment]
I can propose one simple smoothing: compute `y[0] = y[-1] = 0.5 * (y[0] + y[-1])`. Intuitively it will give a desired effect, when the jump is small.

Should I implement this?

**Comment by josef-pkt** [comment]
@nmayorov  Answering that is out of my area.
So far my comments have been from a user perspective, and in analogy to our smoothing spline estimation cases. I never looked at details of the boundary implementation in splines.

> compute `y[0] = y[-1] = 0.5 * (y[0] + y[-1])`

My **guess** would be to leave this up to the user, unless there is a clear set of options for how to "adjust" the boundary observations themselves. Also in terms of the domain/support of the function, only user knows where, at which x, to add artificial observations

**Comment by nmayorov** [comment]
This suggestion was not to improve the algorithm, but rather usability: if `y[0] == y[-1]` it works the same as the conventional algorithm, when the values are different the algorithm still gives a reasonable result. Surely this behavior will be explained in the docstring.

**Comment by josef-pkt** [comment]
My usability thinking was if I see the exception, then I replace my original data by concatenated data, e.g. `y = np.concatenate((y, y[0]))`
Wouldn't produce this the more desired behavior in the `periodic` case? Especially if I want to interpolate existing values, not get a smoothed solution.

**Comment by nmayorov** [comment]
Well then, I'll keep everything as it is, which also makes a perfect sense.

**Comment by ev-br** [comment]
`splrep` et al simply ignore `y[-1]`: https://github.com/scipy/scipy/blob/master/scipy/interpolate/fitpack/percur.f#L38

> x(m) only indicates the length of the period of the spline, i.e per=x(m)-x(1).

and 

> y     : real array of dimension at least (m). before entry, y(i)
> c           must be set to the i-th value of the dependent variable y,
> c           for i=1,2,...,m-1. the element y(m) is not used.

While a choice here is rather arbitrary, consistency is a slight advantage I'd say.

**Comment by josef-pkt** [comment]
Another though:
Your proposed periodic extrapolation sounds good.
Continuing with my example, if I have endpoint concatenated observations in interval [0.01, 24.01], then I could extrapolate to 0.0, and go back to the original clock/hour interval with observations in [0, 24)

**Comment by ev-br** [comment]
Nikolay: `extrapolate="periodic"` sounds reasonable to me. One would need to check that this does not slow down a non-periodic case though.

**Comment by josef-pkt** [comment]
@ev-br  I convinced myself that I like the exception, if I were a user of this function :)
AFAICS, the outcome is the same as in splrep, but this code has an additional check that we have all the points that are needed.
For example, if I forget to add the extra point for splrep, I get an unexpected result and might not realize it because the results look "close enough" with many points.

**Comment by josef-pkt** [comment]
(Aside: I'm noisy here not because I'm a user of the function, but because I still need to figure this out for smoothing splines in statsmodels. Last time I used splines for seasonality, I just used regular splines which have a jump at the bounds instead of being periodic.)

**Comment by nmayorov** [comment]
CubicSpline is in a good shape then?

As for `extrapolate='periodic'`, implementing methods like `integrate` will be somewhat challenging. Still want to try it.

**Comment by ev-br** [comment]
> CubicSpline is in a good shape then?

Mostly yes. I'd also ping the -dev mailing list to see if someone wants to chime in about handling the endpoints and/or `extrapolate="periodic"`.

**Comment by ev-br** [comment]
> regular splines which have a jump at the bounds instead of being periodic.

You are surely aware of `splrep(..., per=True)`?

Also fitpack has `concon` and `cocosp` for convexity constraints, but these do not seem wrapped for python and I've no idea of whether they actually do what they say on the box. But this is way OT for this pull request anyway :-).

**Comment by nmayorov** [comment]
It hung here for a while. Can it be merged @ev-br? Then I will continue with periodic interpolation.

File: scipy/interpolate/_cubic.py [hunk 2]
```diff
@@ -491,34 +583,177 @@ def __init__(self, x, y, axis=0, extrapolate=True):
...
+                else:
+                    raise ValueError("bc_type={} is not allowed.".format(bc))
+            else:
+                try:
+                    deriv_order, deriv_value = bc
+                except Exception:
+                    raise ValueError("A specified derivative value must be "
+                                     "given in the form (order, value).")
+
+                deriv_order = int(deriv_order)
```
Comment (ev-br) [inline scipy/interpolate/_cubic.py]: This is a bit too flexible, too many things can be converted to an integer value of one:

```
In [15]: x = np.arange(8)

In [16]: s1 = CubicSpline(x, x**3, bc_type=("natural", (1.5, 0)))

In [17]: s2 = CubicSpline(x, x**3, bc_type=("natural", (True, 0)))
```

File: scipy/interpolate/_cubic.py [hunk 3]
```diff
@@ -430,18 +471,56 @@ class CubicSpline(PPoly):
...
+    def __init__(self, x, y, axis=0, extrapolate=None, bc_type='not-a-knot'):
         x, y = map(np.asarray, (x, y))
+
         if np.issubdtype(x.dtype, np.complexfloating):
             raise ValueError("`x` must contain real values.")
 
+        if np.issubdtype(y.dtype, np.complexfloating):
+            dtype = complex
+        else:
+            dtype = float
```
Comment (ev-br) [inline scipy/interpolate/_cubic.py]: The dtype selection logic should account for the boundary derivatives, if specified:

```
In [19]: s2 = CubicSpline(x, x**3, bc_type=("natural", (1, 2j)))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-19-40e5338626d5> in <module>()
----> 1 s2 = CubicSpline(x, x**3, bc_type=("natural", (1, 2j)))

/home/br/repos/scipy/build/testenv/lib/python3.5/site-packages/scipy/interpolate/_cubic.py in __init__(self, x, y, axis, extrapolate, bc_type)
    677                     A[1, -1] = 1
    678                     A[-1, -2] = 0
--> 679                     b[-1] = bc_end[1]
    680                 elif bc_end[0] == 2:
    681                     A[1, -1] = 2 * dx[-1]
```

File: scipy/interpolate/_cubic.py (hunk 2 above)
Comment (nmayorov) [inline scipy/interpolate/_cubic.py]: I would just remove `int`, i. e. I don't think it is a good practice to check here for types. Would it be better this way?

File: scipy/interpolate/_cubic.py (hunk 3 above)
Comment (nmayorov) [inline scipy/interpolate/_cubic.py]: Interesting point, I'll try to fix it.

File: scipy/interpolate/_cubic.py [hunk 4]
```diff
@@ -430,18 +471,56 @@ class CubicSpline(PPoly):
...
+    >>> plt.plot(cs(xs)[:,0], cs(xs)[:,1], label='spline')
+    >>> plt.axes().set_aspect('equal')
+    >>> plt.legend(loc='center')
+    >>> plt.show()
+
+    The third example is the interpolation of a polynomial y = x**3 on the
+    interval 0 <= x<= 1. A cubic spline can represent this function exactly.
+    To achieve that we need to specify values and first derivatives at
+    endpoints of the interval. Note that y' = 3 * x**2 and thus y'(0) = 0 and
+    y'(1) = 3.
```
Comment (ev-br) [inline scipy/interpolate/_cubic.py]: Asterisks, possibly also `<=`, should be shielded from Sphinx. I guess a rebase on master should trigger a bright-and-shiny doc build on CircleCI.

File: scipy/interpolate/_cubic.py (hunk 2 above)
Comment (ev-br) [inline scipy/interpolate/_cubic.py]: I would rather `if int(deriv_order) != deriv_order: raise ValueError`. But this is a corner case anyway.

File: scipy/interpolate/_cubic.py [hunk 5]
```diff
@@ -430,18 +471,56 @@ class CubicSpline(PPoly):
...
+    >>> print("ds/dx={:.1f} ds/dy={:.1f}".format(cs(0, 1)[0], cs(0, 1)[1]))
+    ds/dx=0.0 ds/dy=1.0
+    >>> xs = 2 * np.pi * np.linspace(0, 1, 100)
+    >>> plt.figure(figsize=(6.5, 4))
+    >>> plt.plot(np.cos(xs), np.sin(xs), label='true')
+    >>> plt.plot(y[:,0], y[:,1], 'o', label='breakpoints')
+    >>> plt.plot(cs(xs)[:,0], cs(xs)[:,1], label='spline')
+    >>> plt.axes().set_aspect('equal')
+    >>> plt.legend(loc='center')
+    >>> plt.show()
```
Comment (ev-br) [inline scipy/interpolate/_cubic.py]: This is really really minor an non-binding: While you're at it, it'd be good to tweak the example. At least spaces after commas please.

**Comment by ev-br** [comment]
LGTM modulo a couple of nitpicks in trivial corner cases.

File: scipy/interpolate/_cubic.py (hunk 2 above)
Comment (nmayorov) [inline scipy/interpolate/_cubic.py]: There is a check for `deriv_order` being 1 or 2 right below, it should be enough to handle all possible mismatches.

Comment (ev-br) [inline scipy/interpolate/_cubic.py]: Cool then.

File: scipy/interpolate/_cubic.py [hunk 6]
```diff
@@ -392,19 +427,24 @@ class CubicSpline(PPoly):
...
+
+    When 'not-a-knot' boundary conditions is applied to both ends, the
+    resulting spline will be the same as returned by `splrep` (with ``s=0``)
+    and `InterpolatedUnivariateSpline`, but those two methods use a
+    representation in B-spline basis.
+
+    Currently `PPoly` can't correctly evaluate periodic polynomials for
+    arbitrary x. To decrease possible confusion, `extrapolate` is set to False
+    by default when ``bc_type='periodic'``. Fow now you can achieve periodic
+    evaluation for any x as follows::
```
Comment (ev-br) [inline scipy/interpolate/_cubic.py]: So that we don't forget: This paragraph will need to be addressed together with gh-6186

File: scipy/interpolate/_cubic.py (hunk 4 above)
Comment (nmayorov) [inline scipy/interpolate/_cubic.py]: I checked locally and this block is displayed correctly. Looks like Sphinx markdown is not applied for such blocks in Examples section.

**Comment by nmayorov** [comment]
@ev-br, I fixed your remarks. Don't know what's going on with circleci.

**Comment by pv** [comment]
circleci probably wants a rebase to read the correct config file.

**Comment by nmayorov** [comment]
Here is the rendered [docstring](https://circle-artifacts.com/gh/scipy/scipy/10/artifacts/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.interpolate.CubicSpline.html#scipy.interpolate.CubicSpline) to check that it's all right.

**Comment by nmayorov** [comment]
@ev-br @pv can you find time to make a final review (if necessary) and merge? Maybe during the weekend.

**Comment by ev-br** [comment]
Unless there are more comments, I'll hit the green button soon (tm).

**Comment by ev-br** [comment]
OK, let's get this in. Thank you @nmayorov, @matmel @josef-pkt

File: scipy/interpolate/_cubic.py (hunk 4 above)
Comment (ev-br) [inline scipy/interpolate/_cubic.py]: Right. My inclination would still be to typeset code as code.

`x**3` is written with a python syntax, so it should be typeset as code. For the math text we are not really constent, but at any rate the apostrophe in `y'` is not an apostrophe as used in English text, so it should ideally be typeset as maths or code IMO.
